doc/design/block.md

```diff
@@ -0,0 +1,167 @@
+# Design Doc: Use Block in RNNOp, While Op, IfElseOp
+
+In C++ and Java programming language, a block is a lexical structure of source code which is grouped as one line of code.
+
+RNNOp looks like the loop structure in programming languages.
+And similarly, WhileOp and IfElseOp are like loop and conditions respectively.
+So we want to verify if we should have a class Block in PaddlePaddle that works like a pair of curly braces in the loop and condition structures of programming languages.
+
+Blocks do not only group source code, but also narrow the lexical scope of variables so that they do not conflict with variables having the same name used elsewhere in a program.
+
+In Paddle, we need a similar concept called Block to support following scenes:
+
+- define a PaddlePaddle program by writing blocks of codes, which includes the definitions of variables and operators.
+  - `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp` needs Block to help to define sub-block. 
+- help to execute multiple operators, blocks should group operators and runs like a single operator.
+
+## How to use Block
+In `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp`, a sub-block should be used to help to define a sub-block.
+
+Let's start from how a RNNOp is described using Block:
+
+```python
+v = some_op()
+m_boot = some_op()
+
+W = pd.Variable(shape=[20, 20])
+U = pd.Varable(shape=[20, 20])
+
+rnn = create_RNNOp()
+with rnn.stepnet() as net:
+  # declare the input variables that need to be segmented into steps
+  x = net.add_input(v)
+  # declare rnn's memory (state)
+  h = net.add_memory(init=m_boot)
```

Comment:
Who is writing this memory? How to specify what is memorized?

For example, is that `h` memorize `hidden_out` or `act`?

Reply:
done

Reply:
If the current timestamp is `t`, how to get the output from `t-2`. Or more general, `t-n`?

Reply:
may be supported in the future. currently only n-1 is supported

or we can add `pre_memory(n=2)`